File: binary_right_traversal.py
################## Iterative solution
# Time complexity: O(n)
# Space complexity: O(maximum length of level or diameter)
# Worked on leetcode: yes
# Iterative approach: This algorithm uses a queue. Iteratively, we push the root. If the queue is not empty, we keep popping out elements, and for each pop,
# we push the its children. During each push operation. We maintain the size as the length of the queue, and this size determines the number of elements in a
# level. We add the last element of a partiaular level in the output list.

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
        if root == None:
            return []

        queue = []
        queue.append(root)  # parent of root as -1
        output = []

        while (len(queue) > 0):
            size = len(queue)

            for i in range(size):
                node = queue.pop(0)  # get the element

                if i == size - 1:
                    output.append(node.val)

                if node.left != None: queue.append(node.left)
                if node.right != None: queue.append(node.right)

        return output

File: binaryRightTraversal.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :no

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<Integer> result;
    public List<Integer> rightSideView(TreeNode root) {
        result = new ArrayList<Integer>();
        if(root == null)
            return result;
        helper(root, 0);
        return result;   
    }
    
    //traversing left-root-right
    private void helper(TreeNode root, int level){
        //base
        if(root == null) return;
        //logic
        if(result.size() == level)
            result.add(root.val); //new level 
        else
            result.set(level, root.val);//found new RightMost node at that level
        
        helper(root.left, level+1);
        helper(root.right, level+1);
    }
}